File: Codeforces/2176D.py
import sys
input = sys.stdin.readline
# import bisect
from bisect import bisect_left


MOD = 998244353

# def

data = list(map(int, sys.stdin.buffer.read().split()))
it = iter(data)
out_lines = []

# t = int(input())
t = next(it)

for _ in range(t):
    # n, m = map(int, input().split())
    # a = list(map(int, input().split()))
    n = next(it)
    m = next(it)
    a = [next(it) for _ in range(n)]

    fr = [0] * (m+10)
    to = [0] * (m+10)
    out_edges = [[] for _ in range(n+10)]

    for edge in range(m):
        # u, v = map(int, input().split())
        u = next(it)
        v = next(it)
        
        u -= 1; v -= 1
        fr[edge] = u
        to[edge] = v
        out_edges[u].append(edge)

    out_vals = [[] for _ in range(n+10)]
    for i in range(n):
        out_edges_i = out_edges[i]
        out_edges_i.sort(key=lambda x: a[to[x]])
        vals = [a[to[e]] for e in out_edges_i]
        out_vals[i] = vals

    dp = [-1] * (m+10)

    def dp_u_to_v(edge):
        if dp[edge] != -1:
            return dp[edge]

        u = fr[edge]
        v = to[edge]
        w = a[u] + a[v]
        # print(">>> edge:", edge, "from", u, "to", v, "a[u]:", a[u], "a[v]:", a[v], "a[u]+a[v]:", w)

        # begin with from u to v, 1 solution
        ans = 1

        out_edges_v = out_edges[v]
        out_vals_v = out_vals[v]

        idx = bisect_left(out_vals_v, w)
        while idx < len(out_edges_v):
            if out_vals_v[idx] != w:
                break
            ans += dp_u_to_v(out_edges_v[idx])
            # ans %= MOD
            if ans >= MOD:
                ans -= MOD
            idx += 1

        dp[edge] = ans
        return ans
    
    ANS = 0
    for edge in range(m):
        ANS += dp_u_to_v(edge)
        # ANS %= MOD
        if ANS >= MOD:
            ANS -= MOD

    # print(ANS)
    # print(">>>", ANS)
    out_lines.append(str(ANS))
    
print('\n'.join(out_lines))

File: Codeforces/2176D.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long

const ll MOD = 998244353;

vector<ll> a;
vector<int> fr, to;
vector<vector<int>> out_edges;
vector<vector<ll>> out_vals;
vector<ll> dp;

// ll dp_u_to_v(int edge) {
//     if (dp[edge] != -1) return dp[edge];

//     int u = fr[edge];
//     int v = to[edge];
//     ll w = a[u] + a[v];

//     // cout << "u:" << u << " v:" << v << "\t";
//     // cout << "a[u]" << a[u] << " a[v]:" << a[v] << "\t";
//     // cout << "a[u] + a[v]:" << w << "\n";

//     ll ans = 1;

//     auto &out_edges_v = out_edges[v];
//     auto &out_vals_v  = out_vals[v];

//     int idx = (int)(lower_bound(out_vals_v.begin(), out_vals_v.end(), w) - out_vals_v.begin());
//     while (idx < (int)out_edges_v.size()) {
//         if (out_vals_v[idx] != w) break;
//         ans += dp_u_to_v(out_edges_v[idx]);
//         if (ans >= MOD) ans -= MOD;
//         ++idx;
//     }

//     dp[edge] = ans;
//     return ans;
// }

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int t;
    if (!(cin >> t))    return 0;

    while (t --) {
        int n, m;
        cin >> n >> m;

        a.assign(n, 0);
        for (int i = 0; i < n; ++i) {
            cin >> a[i];
        }

        fr.assign(m + 10, 0);
        to.assign(m + 10, 0);
        // out_edges.assign(n + 10, {});
        // out_vals.assign(n + 10, {});
        dp.assign(m + 10, -1);

        for (int edge = 0; edge < m; ++edge) {
            int u, v;
            cin >> u >> v;
            --u;    --v;
            fr[edge] = u;
            to[edge] = v;
            // out_edges[u].push_back(edge);
        }

        // out_edges[i].sort(key=lambda x: a[to[x]])
        // out_vals[i] = [a[to[e]] for e in out_edges_i]
        // for (int i = 0; i < n; ++i) {
        //     auto &out_edges_i = out_edges[i];
        //     sort(out_edges_i.begin(), out_edges_i.end(),
        //          [&](int x, int y) { return a[to[x]] < a[to[y]]; });

        //     auto &vals = out_vals[i];
        //     vals.reserve(out_edges_i.size());
        //     for (int e : out_edges_i) {
        //         vals.push_back(a[to[e]]);
        //     }
        // }

        // ll ANS = 0;
        // for (int edge = 0; edge < m; ++edge) {
        //     ANS += dp_u_to_v(edge);
        //     if (ANS >= MOD)     ANS -= MOD;
        // }

        // cout << ANS << '\n';

        vector<int> ord(m);
        iota(ord.begin(), ord.end(), 0);
        sort(ord.begin(), ord.end(), [&](int e1, int e2) {
            if (a[to[e1]] != a[to[e2]]) return a[to[e1]] > a[to[e2]];
            return fr[e1] < fr[e2];
        });


    }

    return 0;
}